src/window/file_dialog.c

```diff
@@ -78,25 +80,64 @@ static input_box file_name_input = {144, 80, 20, 2, FONT_NORMAL_WHITE, 0, data.t
 static file_type_data saved_game_data = {"sav"};
 static file_type_data scenario_data = {"map"};
 
-static int double_click = 0;
+static int find_first_file_with_prefix(const char *prefix)
+{
+    int len = strlen(prefix);
+    if (len == 0) {
+        return -1;
+    }
+    int left = 0;
+    int right = data.file_list->num_files;
+    while (left < right) {
+        int middle = (left + right) / 2;
+        if (string_compare_case_insensitive_prefix(data.file_list->files[middle], prefix, len) >= 0) {
+            right = middle;
+        } else {
+            left = middle + 1;
+        }
+    }
+    if (string_compare_case_insensitive_prefix(data.file_list->files[left], prefix, len) == 0) {
+        return left;
+    } else {
+        return -1;
+    }
+}
+
+static void scroll_to_typed_text(void)
+{
+    char name_utf8[FILE_NAME_MAX];
+    encoding_to_utf8(data.typed_name, name_utf8, FILE_NAME_MAX, encoding_system_uses_decomposed());
+    int index = find_first_file_with_prefix(name_utf8);
+    if (index >= 0) {
+        scrollbar_reset(&scrollbar, calc_bound(index, 0, data.file_list->num_files - NUM_FILES_IN_VIEW));
+    }
+}
 
 static void init(file_type type, file_dialog_type dialog_type)
 {
     data.type = type;
     data.file_data = type == FILE_TYPE_SCENARIO ? &scenario_data : &saved_game_data;
-    if (strlen(data.file_data->last_loaded_file) == 0) {
-        string_copy(lang_get_string(9, type == FILE_TYPE_SCENARIO ? 7 : 6), data.typed_name, FILE_NAME_MAX);
-        encoding_to_utf8(data.typed_name, data.file_data->last_loaded_file, FILE_NAME_MAX, 0);
-        file_append_extension(data.file_data->last_loaded_file, data.file_data->extension);
-    } else {
+    data.dialog_type = dialog_type;
+
+    data.message_not_exist_start_time = 0;
+    data.double_click = 0;
+    data.focus_button_id = 0;
+
+    if (strlen(data.file_data->last_loaded_file) > 0) {
         encoding_from_utf8(data.file_data->last_loaded_file, data.typed_name, FILE_NAME_MAX);
         file_remove_extension(data.typed_name);
+    } else if (dialog_type == FILE_DIALOG_SAVE) {
+        // Suggest default filename
+        string_copy(lang_get_string(9, type == FILE_TYPE_SCENARIO ? 7 : 6), data.typed_name, FILE_NAME_MAX);
```

Comment:
Just occured to me. Instead of a simple `My Rome`, why not display the name of the city and maybe the date as the default filename?

Reply:
Then the File dialog needs to know about cities and empire maps, and this is not applicable to scenarios in the editor. Besides, "My Rome" is only used if the user has not loaded a file of the same type before. So if you load saved game "Massilia 120AD", then start a new CCK scenario and try to save it, it will still suggest "Massilia 120AD" as file name.
I'd rather keep the "My Rome" (or "Scenario1") than complicate this screen with more logic.